File: app_manager/appmanager/src/main/java/app/editors/manager/ui/views/animation/HeightValueAnimator.java
package app.editors.manager.ui.views.animation;


import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ValueAnimator;
import android.view.View;

public class HeightValueAnimator {

    private static final int DEFAULT_DURATION = 300;
    private static final int MIN_HEIGHT = 0;

    public interface OnAnimationListener {
        void onStart(boolean isShow);
        void onEnd(boolean isShow);
    }

    private final int mDuration;
    private final int mViewHeight;
    private final View mViewAnimation;
    private ValueAnimator mLayoutAnimation;
    private OnAnimationListener mOnAnimationListener;

    public HeightValueAnimator(final View view, final int duration) {
        mViewAnimation = view;
        mViewAnimation.measure(View.MeasureSpec.UNSPECIFIED, View.MeasureSpec.UNSPECIFIED);
        mViewHeight = view.getMeasuredHeight();
        mDuration = duration;
    }

    public HeightValueAnimator(final View view) {
        this(view, DEFAULT_DURATION);
    }

    public void setOnAnimationListener(OnAnimationListener onAnimationListener) {
        mOnAnimationListener = onAnimationListener;
    }

    public void animate(final boolean isShow) {
        final int newVisibility = isShow? View.VISIBLE : View.GONE;
        final int viewVisibility = mViewAnimation.getVisibility();
        if (newVisibility != viewVisibility) {
            if (mLayoutAnimation != null && mLayoutAnimation.isRunning()) {
                mLayoutAnimation.cancel();
            }

            if (isShow) {
                mLayoutAnimation = ValueAnimator.ofInt(MIN_HEIGHT, mViewHeight);
            } else {
                mLayoutAnimation = ValueAnimator.ofInt(mViewHeight, MIN_HEIGHT);
            }

            mLayoutAnimation.addListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationStart(Animator animation) {
                    mViewAnimation.setVisibility(View.INVISIBLE);
                    if (mOnAnimationListener != null) {
                        mOnAnimationListener.onStart(isShow);
                    }
                }

                @Override
                public void onAnimationEnd(Animator animation) {
                    if (isShow) {
                        mViewAnimation.setVisibility(View.VISIBLE);
                    } else {
                        mViewAnimation.setVisibility(View.GONE);
                    }

                    if (mOnAnimationListener != null) {
                        mOnAnimationListener.onEnd(isShow);
                    }
                }
            });

            mLayoutAnimation.setDuration(mDuration);
            mLayoutAnimation.addUpdateListener(animation -> {
                final int value = (int) animation.getAnimatedValue();
                mViewAnimation.getLayoutParams().height = value;
                mViewAnimation.requestLayout();
            });

            mLayoutAnimation.start();
        }
    }

    public void animate() {
        animate(mViewAnimation.getVisibility() == View.GONE);
    }

    public void clear() {
        mOnAnimationListener = null;
        if (mLayoutAnimation != null) {
            mLayoutAnimation.removeAllUpdateListeners();
            mLayoutAnimation.cancel();
        }
    }

}

File: toolkit/libtoolkit/src/main/java/lib/toolkit/base/ui/views/animation/HeightValueAnimator.kt
package lib.toolkit.base.ui.views.animation


import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ValueAnimator
import android.view.View
import android.view.ViewTreeObserver

class HeightValueAnimator(private val view: View, private val duration: Int = DEFAULT_DURATION) {

    companion object {
        private const val DEFAULT_DURATION = 300
        private const val MAX_HEIGHT = 100
        private const val MIN_HEIGHT = 0
    }

    private inner class AnimationAdapter : AnimatorListenerAdapter(), ValueAnimator.AnimatorUpdateListener {

        override fun onAnimationStart(animation: Animator) {
            view.visibility = View.VISIBLE
            onAnimationListener?.onAnimationStart(animation)
        }

        override fun onAnimationEnd(animation: Animator) {
            if (isHideView) {
                view.visibility = if(isShow) View.VISIBLE else View.GONE
            }
            onAnimationListener?.onAnimationEnd(animation)
        }

        override fun onAnimationRepeat(animation: Animator) {
            super.onAnimationRepeat(animation)
            onAnimationListener?.onAnimationRepeat(animation)
        }

        override fun onAnimationCancel(animation: Animator) {
            super.onAnimationCancel(animation)
            onAnimationListener?.onAnimationCancel(animation)
        }

        override fun onAnimationUpdate(animation: ValueAnimator) {
            val value = animation.animatedValue as Int
            view.layoutParams.height = value
            view.requestLayout()
            onAnimationUpdateListener?.onAnimationUpdate(animation)
        }
    }

    var isHideView = true
        private set

    var isShow = false
        private set

    val viewHeight: Int
        get() = view.height

    var onAnimationListener: Animator.AnimatorListener? = null
    var onAnimationUpdateListener: ValueAnimator.AnimatorUpdateListener? = null
    var onLayoutTreeListener: ((View) -> Unit)? = null

    var viewMinHeight = MIN_HEIGHT
    var viewMaxHeight = MAX_HEIGHT
    var viewMeasuredHeight = 0

    private var layoutAnimation: ValueAnimator? = null
    private var animationAdapter: AnimationAdapter = AnimationAdapter()

    init {
        view.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                view.viewTreeObserver.removeOnGlobalLayoutListener(this)
                onLayoutTree()
            }
        })
    }

    protected fun onLayoutTree() {
        viewMeasuredHeight = view.height
        viewMaxHeight = viewMeasuredHeight
        onLayoutTreeListener?.invoke(view)
    }

    fun animate(isShow: Boolean, isHideView: Boolean = true) {
        clear()

        this@HeightValueAnimator.isShow = isShow
        this@HeightValueAnimator.isHideView = isHideView

        layoutAnimation = if (isShow) {
            ValueAnimator.ofInt(viewMinHeight, viewMaxHeight)
        } else {
            ValueAnimator.ofInt(viewMaxHeight, viewMinHeight)
        }.apply {
            duration = duration.toLong()
            addListener(animationAdapter)
            addUpdateListener(animationAdapter)
        }.also {
            it.start()
        }
    }

    fun animate(isShow: Boolean) {
        val newVisibility = if (isShow) View.VISIBLE else View.GONE
        val viewVisibility = view.visibility
        if (newVisibility != viewVisibility) {
            animate(isShow, true)
        }
    }

    fun clear() {
        layoutAnimation?.apply {
            cancel()
            removeAllListeners()
            removeAllUpdateListeners()
        }
    }

}